File: validParenthesis.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> set = new Stack<>();
        char[] ch = s.toCharArray();
        
        for(int i=0; i< ch.length;i++){
            if(ch[i] == '(' || ch[i] == '[' || ch[i] == '{'){
                set.push(ch[i]);
            }else{
                 if(ch[i] == ')'){
                    if(set.isEmpty() || set.pop() != '(') return false;
                 }if(ch[i] == ']'){
                    if(set.isEmpty() || set.pop() != '[') return false;
                 }if(ch[i] == '}'){
                    if(set.isEmpty() || set.pop() != '{') return false;
                 }
            }
        }
        return set.isEmpty();
    }
}

File: valid_parenthesis.cpp
// Time complexity-O(n)
// space complexity O(n);
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this : no

// iterating over the string
// if its an opening bracket push into the stack
// else check it the top corresponds with the top element in the stack before popping it
// if stack is empty after iteration return true

#include<algorithm>
#include<iostream>
#include<vector>
#include<stack>
using namespace std;
class Solution {
public:
    bool isValid(string s) {
        
        if(s.size()== 0){
            return false;
        }
        
        stack<char>st;
        
        for(char c : s){
         
               if( c==']' || c=='}' || c==')'){
               
                if(st.empty()){return false;}
                if( c==']' && st.top()!= '[') return false;
                if( c=='}' && st.top()!= '{') return false;
                if( c==')' && st.top()!= '(' )return false;
                st.pop();  
            }
            else{
                st.push(c);
            }
           
        }
        if(st.empty()){return true;}
        return false;
    }
};